### ISSUE TITLE
Error accessing google drive through APIs and service account

### ISSUE THREAD

**Issue by robyone67** [google/google-api-php-client#1272]
Hello,
I've created a web-site where every user can upload documents that are automatically loaded into a specific google drive space for each user, without needing to enter credentials every time. The gmail user of each user must already exist or I create it for the occasion, but it must be accessible through a graphical interface. Therefore, I use the google drive APIs and create, for each user, a service account and the credentials needed to connect to the existing user to load the files in this drive space.

There is an initial page where the user confirms, once, at the beginning, the project (website)'s access to the user. It runs one time and then never again. This is the code:

//PAGE 1
define('DRIVE_SCOPE', 'https://www.googleapis.com/auth/drive');
$redirect_uri = 'http://' . $_SERVER['HTTP_HOST'] . $_SERVER['PHP_SELF'];
session_start();
$auth_file = "path of client_secret_xxxxxxx.json";
$client = new Google_Client();
$client->setAuthConfig($auth_file);
$client->setRedirectUri($redirect_uri);
$client->setIncludeGrantedScopes(true);
$client->addScope(DRIVE_SCOPE);
$service = new Google_Service_Drive($client);

if (isset($_GET['code'])) {
  $token = $client->fetchAccessTokenWithAuthCode($_GET['code']);
  $client->setAccessToken($token);

  // store in the session also
  $_SESSION['upload_token'] = $token;

  // redirect back to the example
  //header('Location: ' . filter_var($redirect_uri, FILTER_SANITIZE_URL));
  exit("Authentication made!");
}

// set the access token as part of the client
if (!empty($_SESSION['upload_token'])) {
  $client->setAccessToken($_SESSION['upload_token']);
  if ($client->isAccessTokenExpired()) {
    unset($_SESSION['upload_token']);
  }
else {
  $authUrl = $client->createAuthUrl();
}

if (isset($authUrl))
	header('Location: ' . filter_var($authUrl, FILTER_SANITIZE_URL));
`

Then there is a page I call when I want to do an operation of any kind: upload, download, read, etc. This is the authentication code that runs before the operation:

`//PAGE 2
define('DRIVE_SCOPE', 'https://www.googleapis.com/auth/drive');
$email = “gmail account of user”
$client = new Google_Client();

// set the scope(s) that will be used
$client->addScope(DRIVE_SCOPE);

// this is needed only if you need to perform
// domain-wide admin actions, and this must be
// an admin account on the domain; it is not 
// necessary in your example but provided for others
$client->setSubject($email);

// set the authorization configuration using the 2.0 style
$auth_file = "path of service-account.json";
$client->setAuthConfig($auth_file);
`

**Until recently it worked regularly, I created several users like this! It still works for users already created!**

For some time now, when I create a new user:

- when I run the authentication page (page 1), I get an error:
Error: invalid_scope
You don't have permission to access some scopes. Your project is trying to access scopes that need to go through the verification process. {invalid = [https://www.googleapis.com/auth/drive]} If you need to use one of these scopes, submit a verification request.
If I join the "Risky Apps" group, this problem resolves, but not what follows.

- When I try to do any operation on the drive, page 2 authentication does not return any error, but when it comes to perform the specific operation, even a simple "file list", I get the following error:
{"Error": "unauthorized_client", "error_description": "Client is unauthorized to retrieve access tokens using this method." }
In short, I can not do anything on the gmail user drive.
I found that if, on page2, comment the line "$client->setSubject ($email);" or I put the service account mail address, the procedure works, but it works in the service account drive, not in the gmail account.

Can someone help me? I would like the procedure to work and work in an existing gmail account, as it did before! Definitely Google has introduced some changes ... Is this something that can be solved?
Thank you.

**Comment by robyone67** [comment]
**No one can help me? Basically, I found out that I can no longer use google drive apis through a service account, "impersonating" an existing user! I've found several similar posts here, but nobody deals with this problem. Does anyone have the same problem?**

**Comment by bryannielsen** [comment]
I saw a bunch of "unauthorized_client" errors recently and I was using the latest version `2.2.0`.  After downgrading to `2.1.3` everything is working fine with impersonating users.

**Comment by mattwhisenhunt** [comment]
@robyone67 sorry for the delay on this one. Did you try downgrading to resolve this issue?

**Comment by mattwhisenhunt** [comment]
This issue may no longer be relevant due to its age. If that is not the case let us know and we will re-open.
